Determine final paper decision based on (default) decision matrix

Merseleo opened the pull request:
Closes #345 

## What I have made

<!-- write down what changes have been made, what was removed/added/changed -->
<!-- e.g. I added a service, which does x -->

- I added a meaningful default decision matrix (based on two reviewers) to the user settings that is used when a new project is created
- I extended the `createReview` service function, so that the paper decision is not updated based on a simple summation function anymore, but either
  - set to  `DECLINED` when the decision is declined and justified with at least one hard exclusion criterion
  - check the paper decision by pattern match the current reviews to the review decision matrix of the project
- I extended and refactored the `CreateReviewTest` to be up-to-date with the new paper decision update procedure 

## Checklist

Either tick or cross out the items that do not apply (using \~\~example text\~\~) and give a reason why the item does
not apply.

- [x] I have updated the documentation accordingly and commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] (for reviewer) I have checked the implementation against the requirements

Commit:
feat: add meaningful default review decision matrix for user with encoded review patterns

Commit:
refactor: update paper decision after review based on review decision matrix instead of summation of review decisions.

Additionally, introduced a helper function in `DataBuilder` to build a decision matrix in tests.

Commit:
feat: decline a project paper if at least one hard exclusion criterion is selected during review

Commit:
test: add missing tests for the 'createReview' call

These include testing that the paper is finally decided when a hard exclusion criterion is selected or that the arbiter decision is handled correctly.

As the 'CreateReviewTest' file has become quite large with the new tests, it was refactored and duplicate / similar code extracted.

Enhanced test readability and modularity by replacing the generic `mockHappyPathUntil` with customized `mockCreateReview` utility, allowing better control of specific test scenarios and stopping points.

Merseleo commented:
Note for the reviewer: I would be glad, if you can also manually test this update. The logic of creating reviews has become quite complex and it should work, but maybe I forgot a scenario that you can discover when manually testing it.

Merseleo committed:
chore: enable TODO marker detection in detekt configuration

luca-schlecker commented on src/main/kotlin/se/uulm/snowballr/backend/service/ReviewService.kt:
```diff
@@ -96,49 +101,51 @@ class ReviewService(
-            projectPaperRepo.updateProjectPaperDecision(projectPaperId, PaperDecision.PAPER_DECISION_IN_REVIEW)
-            return
+    private fun determinePaperDecision(reviews: List<Review>, decisionMatrix: ReviewDecisionMatrix): PaperDecision {
+        if (reviews.size < decisionMatrix.numberOfReviewers) {
```
There is a check for `reviews.size == 0` missing, which should set the decision to `PAPER_DECISION_UNREVIEWED`, as the function documentation states.

luca-schlecker replied:
This is solved now :+1:

luca-schlecker commented on src/main/kotlin/se/uulm/snowballr/backend/table/UserTable.kt:
```diff
@@ -7,21 +7,71 @@ import org.jetbrains.exposed.sql.ResultRow
+)
+private val MAYBE_MAYBE_PATTERN = patternOf(
+    ReviewDecision.REVIEW_DECISION_MAYBE to 1L,
+    ReviewDecision.REVIEW_DECISION_MAYBE to 1L,
```
Should be merged to `2L`

luca-schlecker commented on src/test/kotlin/se/uulm/snowballr/backend/DataBuilder.kt:
```diff
@@ -269,4 +270,36 @@ object DataBuilder {
+    )
+    private val MAYBE_MAYBE_PATTERN = patternOf(
+        ReviewDecision.REVIEW_DECISION_MAYBE to 1L,
+        ReviewDecision.REVIEW_DECISION_MAYBE to 1L,
```
Same here, should be `2L`

luca-schlecker reviewed (changes requested):
Only three small things I've noticed 👍 The implementation seems good for now

Merseleo committed:
refactor: combined MAYBE_MAYBE_PATTERN decisions into one map

Merseleo committed:
fix: explicitly catch the case where the paper decision should be determined for an unreviewed paper

 Although this is actually not possible in the current implementation, we add this case (without test, as the test is not easily possible), because this private helper function could be used later for different purposes.

luca-schlecker reviewed (dismissed):
All issues I've noticed are addressed now. Looking good to me 👍

luca-schlecker reviewed (dismissed):
Looking good to me 👍

luca-schlecker reviewed (changes requested):
Pipeline's are not running
